Add section to GOI docs with details on mapping genes (#284)
AlexsLemonade/scpca-downstream-analyses: optional-goi-analysis/README.md

3. The genes of interest list, stored as a tab-separated value (TSV) file.
This file should contain at least one column named `gene_id` with the relevant gene identifiers, and can optionally contain an additional column named `gene_set` that denotes the gene set that each gene identifier belongs to (see an example of this genes of interest file [here](../example-data/goi-lists/sample01_goi_list.tsv)).

## Gene mapping

> The first step in the workflow is to ensure that the input gene identifiers used in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 

Reviewer: ```suggestion
The first step in the workflow is to ensure that the input gene identifiers provided in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
```

The `SingleCellExperiment` objects that are returned by the core workflow will contain gene names as Ensembl ids.
If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects, then mapping must be performed.
The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.

To run the gene mapping step, you can run the below command while using the `--config` flag to tailor the genes of interest mapping parameters mentioned in the [gene mapping parameters section above](#gene-mapping-parameters):

```
snakemake --snakefile goi.snakefile \ 
  --cores 2 \
  --use-conda \
  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>"

> ```

Reviewer: I actually think we should change this to show an example of running with mapping and running without mapping.

### Gene mapping parameters

The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for the parameters required for mapping the gene identifiers in the genes of interest workflow.

> It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 

Reviewer: ```suggestion
These parameters will only be used if `perform_mapping` is set to `TRUE`.
It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 
```

The following gene mapping parameters found in the `config/goi_config.yaml` file can be optionally modified:

| Parameter        | Description | Default value |
|------------------|-------------|---------------|
| `organism` | the organism associated with the provided genes and data | "Homo sapiens" |

> | `sce_rownames_identifier` | the type of gene identifiers found in the rownames of the SingleCellExperiment object; note that this parameter should not be changed unless the output of the core analysis workflow has been altered in some way prior to running this module | "ENSEMBL" |

Reviewer: ```suggestion
| `sce_rownames_identifier` | the type of gene identifiers found in the rownames of the `SingleCellExperiment` object; note that this parameter should not be changed unless the output of the core analysis workflow has been altered in some way prior to running this module | "ENSEMBL" |
```

> | `perform_mapping` | a binary value indicating whether or not to perform gene identifier mapping | `TRUE` |

Reviewer: Can you move this to be first in the list.

| `multi_mappings` | how to handle multiple gene identifier mappings when `perform_mapping` is `TRUE` | "list" |

|[View Genes of Interest Config File](../config/goi_config.yaml)|
|---|

> ## Parameters and config file

Reviewer: Sorry for the confusion, I meant to just move the gene mapping before the gene mapping parameters not before all the parameters. This should be moved back above to be before `Gene mapping`. 
Also before you make your next commits and move stuff around don't forget to update the TOC.

Author: Ah oops, got it!!

  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>"
```

## Gene mapping

The first step in the workflow is to ensure that the input gene identifiers provided in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
The `SingleCellExperiment` objects that are returned by the core workflow will contain gene names as Ensembl ids.
If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects, then mapping must be performed.
The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.

> To run the gene mapping step, you can run the below command while using the `--config` flag to tailor the genes of interest mapping parameters mentioned in the [gene mapping parameters section above](#gene-mapping-parameters):

Reviewer: ```suggestion
To run the workflow with the gene mapping step, you will only need to provide the required parameters mentioned in the [project-specific parameters section](#project-specific-parameters).
```
I think we can simplify this and just have another example here of how to run it with mapping, without modifying anything differently. I was thinking about how we show different examples throughout `scpca-nf` and sometimes we have the same command a few times, and I think that's okay because people will read different sections.

```
snakemake --snakefile goi.snakefile \ 
  --cores 2 \
  --use-conda \
  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \

>   provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>" \
>   perform_mapping=TRUE \
>   organism="<NAME OF ORGANISM ASSOCIATED WITH THE PROVIDED GOI>" \
>   sce_rownames_identifier="<TYPE OF GENE IDENTIFIER IN SCE ROWNAMES>" \
>   multi_mappings="list"

Reviewer: ```suggestion
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>"
```
Following my above comment I think we can simplify this.

```

To run the workflow without the gene mapping step, you can run the below command:

```
snakemake --snakefile goi.snakefile \ 
  --cores 2 \
  --use-conda \
  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
  goi_list="<RELATIVE PATH TO YOUR GENES OF INTEREST LIST>" \
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN GOI LIST>" \
  perform_mapping=FALSE
```

### Gene mapping parameters

> The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for the parameters required for mapping the gene identifiers in the genes of interest workflow.

Reviewer: ```suggestion
The [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, sets the defaults for additional parameters required for mapping the gene identifiers in the genes of interest workflow.
```

```

### Gene mapping parameters

The additional [configuration file](https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html), `config/goi_config.yaml`, also sets the defaults for the parameters required for mapping the gene identifiers in the genes of interest workflow.
It is **not required** to alter these parameters to run the workflow, but if you would like to modify the gene identifier mapping, you can do so by changing these parameters. 

> The following gene mapping parameters found in the `config/goi_config.yaml` file can be optionally modified:

Author: Good point!

|[View Genes of Interest Config File](../config/goi_config.yaml)|
|---|

> ## Gene mapping

Reviewer: I actually think you need to move this up above the section you have on parameters and then have the genes of interest parameters within this section. Having the parameters first is confusing because you have introduced them with little context about what mapping actually is.

> If the input gene identifiers do not match the gene identifiers used in the `SingleCellExperiment` object used as input, then input gene identifiers must first be mapped to a specified type.

Reviewer: ```suggestion
The first step in the workflow is to ensure that the input gene identifiers used in the genes of interest list match the gene identifiers present in the `SingleCellExperiment` object. 
The `SingleCellExperiment` objects that are returned by the core workflow will contain gene names as Ensembl ids. 
```

So here I'm suggesting some ways to rephrase this paragraph to make sure we explicitly state what it means to map and why we need to map. We want to be sure we are clear with the fact that the output from the core workflow contains Ensembl Ids so if you input anything else you need to map.

> When the `perform_mapping` flag is set to `TRUE`, users can map the provided gene identifiers to the identifiers used to label the rownames of the `SingleCellExperiment` object.

Reviewer: ```suggestion
If the provided genes of interest list contains another identifier type (e.g., gene symbol, Entrez id) that does not match the gene names present in the `SingleCellExperiment` objects then mapping must be performed.
The `perform_mapping` flag is used to indicate whether or not gene mapping will be performed and by default is set to `TRUE`.
```

To run the gene mapping step, you can run the below command while using the `--config` flag to tailor the genes of interest mapping parameters mentioned in the [gene mapping parameters section above](#gene-mapping-parameters):

```
snakemake --snakefile goi.snakefile \ 
  --cores 2 \
  --use-conda \
  --config results_dir="<RELATIVE PATH TO RESULTS DIRECTORY>" \
  project_metadata="<RELATIVE PATH TO YOUR PROJECT METADATA TSV>" \
  goi_list="<RELATIVE PATH TO YOUR GOI LIST>" \
  organism="<NAME OF ORGANISM ASSOCIATED WITH THE PROVIDED GOI>" \
  provided_identifier="<TYPE OF GENE IDENTIFIER USED IN PROVIDED GOI LIST>" \
  sce_rownames_identifier="<TYPE OF GENE IDENTIFIER IN ROWNAMES OF INPUT SCE OBJECT>" \

>   perform_mapping=TRUE \

Author: Another great point! 👍

| `project_metadata` | relative path to your specific project metadata TSV file (use the same `project_metadata` used in the prerequisite core workflow) |
| `goi_list` | the file path to a tsv file containing the list of genes that are of interest; the default file path is `example-data/goi-lists/example_goi_list.tsv` |
| `provided_identifier` | the type of gene identifiers used to populate the genes of interest list; example values that can implemented here include `"ENSEMBL"`, `"ENTREZID"`, `"SYMBOL"` -- where `"SYMBOL"` is the default (see more keytypes [here](https://jorainer.github.io/ensembldb/reference/EnsDb-AnnotationDbi.html)) |

> | `overwrite` | a binary value indicating whether or not to overwrite existing output files | `TRUE` |

Reviewer: Here you have the default value for `overwrite` but do not have a separate column for the other rows. Can you add default as a separate column for the entire table.
